include/CObjectCache.h

```diff
@@ -70,191 +122,433 @@ namespace impl
-    }
+    public:
+        using Base::Base;
 
-    //! Returns true if had to insert
-    inline bool swapValue(const K& _key, T* _val)
+        inline bool insert(const typename Base::KeyType_impl& _key, typename Base::ValueType_impl _val)
+        {
+            INSERT_IMPL_ASSOC
+        }
+
+        inline typename Base::RangeType findRange(const typename Base::KeyType_impl& _key)
+        {
+            return Base::m_container.equal_range(_key);
+        }
+        inline typename Base::ConstRangeType findRange(const typename Base::KeyType_impl& _key) const
+        {
+            return Base::m_container.equal_range(_key);
+        }
+    };
+    template<
+        template<typename...> class ContainerT_T,
+        typename T, //value type for container
+        typename ...K //optionally key type for std::map/std::unordered_map
+    >
+    struct CMultiObjectCacheBase<false, ContainerT_T, T, K...> : public CObjectCacheBase<ContainerT_T, T, K...>, public CMultiCache_tag
     {
-        greet(_val); //grab before drop
+    private:
+        using Base = CObjectCacheBase<ContainerT_T, T, K...>;
+
+    public:
+        using Base::Base;
 
-        auto it = find(_key);
-        if (it != std::end(m_container))
+        inline bool insert(const typename Base::KeyType_impl& _key, typename Base::ValueType_impl _val)
         {
-            dispose(it->second);
-            it->second = _val;
-            return false;
+            INSERT_IMPL_VEC
         }
 
-        m_container.insert(it, {_key, _val});
-        return true;
-    }
+    private:
+        template<typename RngType>
+        inline RngType findRange_internal(const typename Base::KeyType_impl& _key)
+        {
+            auto cmpf = [](const typename Base::PairType_impl& _a, const typename Base::PairType_impl& _b) -> bool {return _a.first < _b.first; };
+            typename Base::PairType_impl lookingFor{_key, nullptr};
+
+            RngType range;
+            range.first = std::lower_bound(std::begin(Base::m_container), std::end(Base::m_container), lookingFor, cmpf);
+            if (range.first == std::end(Base::m_container) || _key < range.first->first)
+            {
+                range.second = range.first;
+                return range;
+            }
+            range.second = std::upper_bound(range.first, std::end(Base::m_container), lookingFor, cmpf);
+            return range;
+        }
 
-    //! Returns false if item with the _key did not exist in the cache and a nullptr was inserted for it
-    inline bool getByKeyOrReserve(T** _outval, const K& _key)
+    public:
+        inline typename Base::RangeType findRange(const typename Base::KeyType_impl& _key)
+        {
+            return findRange_internal<typename Base::RangeType>(_key);
+        }
+        inline typename Base::ConstRangeType findRange(const typename Base::KeyType_impl& _key) const
+        {
+            return findRange_internal<typename Base::ConstRangeType>(_key);
```

Comment:
how does your compiler like the fact that you are calling `findRange_internal`, a function without a `const` qualifier?

Because GCC sure-as-hell won't like it.

Reply:
although if you made the internal function a `static` and take the container as a parameter then you'd be safe

Reply:
That's a bug. Compiler didn't compile the const variant of function and so didn't give an error while testing. I'll fix as you proposed with static